src/SignalR/common/Shared/MessageBuffer.cs

```diff
@@ -0,0 +1,341 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Internal;
+using Microsoft.AspNetCore.SignalR.Protocol;
+
+namespace Microsoft.AspNetCore.SignalR.Internal;
+
+internal sealed class MessageBuffer : IDisposable
+{
+    private static readonly TaskCompletionSource<FlushResult> _completedTCS = new TaskCompletionSource<FlushResult>();
+
+    private readonly (SerializedHubMessage? Message, long SequenceId)[] _buffer;
+    private readonly ConnectionContext _connection;
+    private readonly IHubProtocol _protocol;
+    private readonly AckMessage _ackMessage = new(0);
+    private readonly SequenceMessage _sequenceMessage = new(0);
+    private readonly Channel<int> _waitForAck = Channel.CreateBounded<int>(new BoundedChannelOptions(1) { FullMode = BoundedChannelFullMode.DropOldest });
+
+#if NET8_0_OR_GREATER
+    private readonly PeriodicTimer _timer = new(TimeSpan.FromSeconds(1));
+#else
+    private readonly TimerAwaitable _timer = new(TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(1));
+#endif
+    private readonly SemaphoreSlim _writeLock = new(1, 1);
+
+    private int _bufferIndex;
+    private long _totalMessageCount;
+    private bool _waitForSequenceMessage;
+
+    // Message IDs start at 1 and always increment by 1
+    private long _currentReceivingSequenceId = 1;
+    private long _latestReceivedSequenceId = long.MinValue;
+    private long _lastAckedId = long.MinValue;
+
+    private TaskCompletionSource<FlushResult> _resend = _completedTCS;
+
+    private object Lock => _buffer;
+
+    static MessageBuffer()
+    {
+        _completedTCS.SetResult(new());
+    }
+
+    // TODO: pass in limits
+    public MessageBuffer(ConnectionContext connection, IHubProtocol protocol)
+    {
+        // Arbitrary size, we can figure out defaults and configurability later
+        const int bufferSize = 10;
```

Comment:
I think the primary or only limit should be based on the total size of the serialized buffer in bytes. 10 messages before observing backpressure could be really bad for any app that sends a lot of small messages over high latency connections.

I think the default should probably be on the order of 100 KB or 1 MB on the server (similar to SocketTransportOptions.MaxReadBufferSize and Http2Limits InitialConnectionWindowSize are both 1 MB) and more on the client to avoid unnecessary backpressure.

On the server, backpressure could slow down some server logic and punish faster connections in the same group because of one slow connection. This is already an issue with socket backpressure, but socket writes usually involve large amounts of buffering at lower layers, and this is all app level.

This wouldn't count for the fact that some of these serialized messages have multiple references, but I think it's okay to overcount for this. At least it's simple to explain. This also wouldn't account for the size of the slots in the collection referencing the SerializedHubMessage, but we can probably mitigate that with smart collection design. Worst case, we could have a separate limit for the number of buffered hub messages, but it should also be large by default. 10 feels too small. 1000 seems more reasonable, but this should definitely be discussed as part of the threat modeling.